### DISCUSSION TITLE
WithObservationTracking with Charts

### DISCUSSION THREAD

**Discussion by Ilya-Floww** [pointfreeco/swift-perception#78] [Q&A] [answered]
Hi, I've run into a problem recently where I am trying to access some state(marked as ObservableState) inside of a ChartContentBuilder closure. 

I am getting a runtime warning letting me know that I need to wrap the content with WithPerceptionTracking however I have noticed that there is no initialiser for content that conforms to ChartContent.

I have _naively_ tried to add my own extension with an init where Content: ChartContent like so:

```swift
extension WithPerceptionTracking: ChartContent where Content: ChartContent {
```

However, the content property of the WithPerceptionTracking struct is not public so that did not work.

Does anyone have a workaround for this or knows why ChartContent has not been included already as part of the library?

Any help would be appreciated. 

For reference my code looks something like this.

```swift

struct MyView: View {
  let store: ...
  var body: some View {
    Chart(store.someObservableStateProperty) {
      // Some code that access another observable state property inside the store.
    }
  }
}

```

I have tried wrapping the whole chart with WithObservationTracking but that did not help.

**Comment by mbrandonw** [comment, marked as answer]
> Does anyone have a workaround for this or knows why ChartContent has not been included already as part of the library?

Hi @Ilya-Floww, thanks for bringing this up! There is no real reason. We just don't use charts ourselves and so didn't even know we needed to provide support.

We would happily take a PR that supports `ChartContent`. Would you have the time to give that a shot?

**Reply by Ilya-Floww** [reply]
Hi @mbrandonw, that makes sense thanks for the quick reply. 

Sure, I'd be happy to open up a PR to add support for `ChartContent`.

**Comment by MartinStrambach** [comment]
Hi guys, I enabled `SWIFT_UPCOMING_FEATURE_MEMBER_IMPORT_VISIBILITY` in my project which still runs on iOS 16 and uses `WithPerceptionTracking`. I get this error:
`Result builder 'ChartContentBuilder' does not implement any 'buildBlock' or a combination of 'buildPartialBlock(first:)' and 'buildPartialBlock(accumulated:next:)' with sufficient availability for this call site`
I don't use Charts framework in my project. Do you have an idea how to go around this issue?

**Reply by Brett-Best** [reply]
I think this is an issue with Swift itself, to workaround it though you just need to add `import Charts`.

**Reply by simba909** [reply]
Late reply but this issue does not appear when compiling with Xcode 26 / Swift 6.2 (even without importing `Charts`).
